File: ios/Drivemetadata.m
#import "Drivemetadata.h"
#import <React/RCTBridgeModule.h>
@import DriveMetaDataiOSSDK;


@implementation Drivemetadata

RCT_EXPORT_MODULE()

// sdk init
RCT_EXPORT_METHOD(sdkInit:(NSInteger)clientID
                  withToken:(NSString *)token
                  withAppId:(NSInteger)appId
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject) {
  // Validate parameters
  if (clientID <= 0 || appId <= 0 || token.length == 0) {
    reject(@"INVALID_PARAMETERS", @"Client ID, token, or app ID is invalid", nil);
    return;
  }

  // Initialize SDK
  @try {
    [DriveMetaData initializeSharedWithClientId:clientID clientToken:token clientAppId:appId];

    resolve(@"DMD SDK Init Successfully");
  } @catch (NSException *exception) {
    reject(@"SDK_INIT_FAILED", @"Failed to initialize SDK", [NSError errorWithDomain:@"DriveMetaDataError" code:500 userInfo:@{NSLocalizedDescriptionKey: exception.reason}]);
  }
}

// Modified method to include a return type
RCT_EXPORT_METHOD(enableIdfa:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  @try {
    NSString *response = [[DriveMetaData shared] requestIDFA]; // Ensure `enableIdfa` exists in SDK
    NSLog(@"enableIdfa   - %@", response);
    resolve(response);
  } @catch (NSException *exception) {
    NSLog(@"Failed to enable IDFA: %@", exception.reason);
    reject(@"IDFA_INIT_FAILED", @"Failed to Get IDFA ID", [NSError errorWithDomain:@"DriveMetaDataError" code:500 userInfo:@{NSLocalizedDescriptionKey: exception.reason}]);  }
}

// sends the data to dmd server

RCT_EXPORT_METHOD(sendTags:(NSDictionary *)data) {
  if (data) {
    
    
    [[DriveMetaData shared] sendTagsWithTags:data eventType:@"delete" completion:^(id response) {
        NSLog(@"Received response: %@", response);
    }];

  } else {
    NSLog(@"Data is null");
  }
}



//handle deeplink

RCT_REMAP_METHOD(getBackgroundData,
                 url:(NSString *)url
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  if (url == nil) {
    reject(@"invalid_url", @"URL cannot be null", nil);
    return;
  }

  @try {
    NSLog(@"Fetching Background Data for URL: %@", url);
    
    [[DriveMetaData shared] getBackgroundDataWithUri:@"https://deep.drivemetadata.com/1mdky" callback:^(NSString * _Nullable jsonString, NSError * _Nullable error) {
        if (error) {
            // Handle error
          reject(@"Error", error.localizedDescription, nil);

        } else if (jsonString) {
          resolve(jsonString);

            // Handle success, use jsonString
            NSLog(@"Amit Kumar Received data: %@", jsonString);
        }
    }];
  } @catch (NSException *exception) {
    reject(@"exception", exception.reason, nil);
  }
}








//E695AE7D-C962-4EB2-9572-665C35422976
RCT_EXPORT_METHOD(appDetails:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    NSString *response = [[DriveMetaData shared] appDetails];
    NSLog(@"appDetails title: %@", response);
    resolve(response); // Resolve the promise with the response
  }
  @catch (NSException *exception) {
    reject(@"app_details_error", @"Failed to fetch app details", nil); // Reject the promise on error
  }
}


RCT_EXPORT_METHOD(deviceDetails:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    NSString *response = [[DriveMetaData shared] deviceDetails];
    NSLog(@"deviceDetails title: %@", response);
    resolve(response); // Resolve the promise with the response
  }
  @catch (NSException *exception) {
    reject(@"app_details_error", @"Failed to fetch app details", nil); // Reject the promise on error
  }
}




+ (BOOL)requiresMainQueueSetup
{
  return NO;
}

@end

File: ios/Drivemetadata.swift
import DriveMetaDataiOSSDK
import AdSupport
import AppTrackingTransparency


@objc(Drivemetadata)
class Drivemetadata: NSObject {
  
  
  
  @objc(sdkInit:withToken:withAppId:withResolver:withRejecter:)
  func sdkInit(clientID: Int, clientToken: String, appId: Int, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
    DriveMetaData.initializeShared(clientId: clientID, clientToken: clientToken, clientAppId: appId)
    
    // Return success message or any needed response
    resolve("SDK initialized successfully")
  }
  
  // send Tags
  
  @objc(sendTags:withEventType:withResolver:withRejecter:)
  func sendTags(tags: NSDictionary, eventType: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    if let tagsDict = tags as? [String: Any] {
      DriveMetaData.shared?.sendTags(tags: tagsDict, eventType: eventType) { response in
        print("Received response: \(response)") // `response` is a non-optional String
        resolve(response)  // Resolve directly
      }
    } else {
      let error = NSError(domain: "DriveMetadata", code: 400, userInfo: [NSLocalizedDescriptionKey: "Invalid tags dictionary"])
      reject("INVALID_INPUT", "Tags parameter is not a valid dictionary", error)
    }
  }
  
  // deeplink
  
  @objc(getBackgroundData:withResolver:withRejecter:)
  func getBackgroundData(url: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    if let url = URL(string: url) {
      DriveMetaData.shared?.getBackgroundData(uri: url) { (jsonString, error) in
        if let error = error {
          reject("BACKGROUND_DATA_ERROR", error.localizedDescription, error)
        } else if let jsonString = jsonString {
          resolve(jsonString)
        } else {
          reject("UNKNOWN_ERROR", "No data received", nil)
        }
      }
      
    } else {
      print("Invalid URL")
    }
  }
  @objc(requestIDFA:rejecter:)
  func requestIDFA(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      if #available(iOS 14, *) {
        ATTrackingManager.requestTrackingAuthorization { status in
          switch status {
          case .authorized:
            let idfa = ASIdentifierManager.shared().advertisingIdentifier.uuidString
            resolve(idfa)
          case .denied, .restricted, .notDetermined:
            resolve(nil) // Return `nil` if not authorized
          @unknown default:
            resolve(nil)
          }
        }
      } else {
        let idfa = ASIdentifierManager.shared().advertisingIdentifier.uuidString
        resolve(idfa)
      }
    }
  }
  
}
